packages/connection-manager/src/ConnectionQueue.js

```diff
@@ -0,0 +1,105 @@
+import { List } from "immutable";
+import ConnectionQueueItem from "./ConnectionQueueItem";
+
+/**
+ * Connection Queue
+ *
+ * The queue is prioritized to enable dynamic connection handling and
+ * best utilize the bandwidth available.
+ *
+ * This impl. wraps `Immutable.List` instead of extending `List` due to
+ * https://github.com/facebook/immutable-js/issues/301
+ */
+export default class ConnectionQueue {
+  /**
+   * Initializes an instance of ConnectionQueue
+   * @param {Set} connections – list of connections
+   */
+  constructor(connections = List()) {
+    if (!List.isList(connections)) {
+      throw new Error(
+        "Given list of connection must to be an instance of List."
+      );
+    }
+
+    /**
+     * @property {List}
+     * @name ConnectionQueue#connections
+     */
+    Object.defineProperty(this, "connections", { value: connections });
+  }
+
+  /**
+   * Return current size of queue                      ´
+   * @return {int} size of queue
+   */
+  get size() {
+    return this.connections.size;
+  }
+
+  /**
+   * returns the first connection
+   * @return {AbstractConnection} - first connection
+   */
+  first() {
+    if (this.connections.first() === undefined) {
+      throw new Error("Cant get first() from empty Queue");
+    }
+
+    return this.connections.first().connection;
+  }
+
+  /**
+   * returns a new queue without the first connection
+   * @return {ConnectionQueue} – ConnectionQueue without first connection
+   */
+  shift() {
+    return new ConnectionQueue(this.connections.shift());
+  }
+
+  /**
+   * Adds given connection with given (or default) priority to queue
+   * @param {AbstractConnection} connection – Connection to be added to queue
+   * @param {Integer} priority – priority of connection
+   * @return {ConnectionQueue} - ConnectionQueue containing the added connection
+   */
+  enqueue(connection, priority) {
+    return new ConnectionQueue(
+      this.connections
+        .push(new ConnectionQueueItem(connection, priority))
+        .sort((a, b) => {
+          const delta = b.priority - a.priority;
+
+          return delta / Math.abs(delta || 1);
```

Comment:
Let's enforce sensible bounds on priority in `ConnectionQueue` and remove this line.

Reply:
How does this relate to priority bounds? It the same as the following lines and what we do in [`ServiceTableUtil` ](https://github.com/dcos/dcos-ui/blob/master/plugins/services/src/js/utils/ServiceTableUtil.js#L11)

```js
.sort((a, b) => {
  if(a. priority > b. priority){
    return -1
  }

  if(a. priority < b. priority){
    return 1
  }

  return 0
}
```

Reply:
Enforcing priority to be an integer means we don't have to think about bizarre floating point arithmetic and I think it's much cleaner - especially because we specified the type as integer. We already validate that priority should not be a sub-zero value. Then `return b.priority - a.priority;` is sufficient. We do not need to scale the values to return `-1`, `1` or `0` specifically. If I have missed something, please help me understand why this last line is needed.